calyx-opt/src/passes/metadata_table_gen.rs

```diff
@@ -0,0 +1,110 @@
+use crate::traversal::{Action, Named, VisResult, Visitor};
+use calyx_ir::Id;
+use calyx_utils::WithPos;
+use linked_hash_map::LinkedHashMap;
+use std::fmt;
+use std::path::PathBuf;
+
+/// Metadata stores a Map between each group name and data used in the metadata table (specified in PR #2022)
+#[derive(Default)]
+pub struct Metadata {
+    groups: LinkedHashMap<Id, (usize, PathBuf)>,
+}
+
+impl Metadata {
+    /// Create an empty metadata table
+    pub fn new() -> Self {
+        Metadata {
+            groups: LinkedHashMap::new(),
+        }
+    }
+
+    /// Add a new entry to the metadata table
+    fn add_entry(&mut self, name: Id, line: usize, path: PathBuf) -> &mut Self {
+        let ins = self.groups.insert(name, (line, path));
+        match ins {
+            None => self,
+            Some(_v) => {
+                println!("Found two of same group name");
+                self
+            }
+        }
+    }
+}
+
+impl fmt::Display for Metadata {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let grps = &self.groups;
+        let mut text = String::new();
+
+        for (name, (line_num, file)) in grps {
+            let file = file.to_str();
+            let file = match file {
+                None => "x",
+                Some(f) => f,
+            };
+            text.push_str(format!("    {name}: {file} {line_num}\n").as_str());
```

Comment:
instead of buffering up the write in a string, you should replace this with
```rust
writeln!("    {name}: {file} {line_num}")?;
```

And just return `Ok(())` at the end